### ISSUE TITLE
[steamshovel] bubbles wrong color-code (Trac #1457)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1449]
Bubbles have the wrong color-code for hits long after the initial hit, which algorithmically currently determines the color of that specific DOM for all times.

Migrated from https://code.icecube.wisc.edu/ticket/1457
```json
{
    "status": "closed", 
    "changetime": "2016-06-21T14:51:48", 
    "description": "Bubbles have the wrong color-code for hits long after the initial hit, which algorithmically currently determines the color of that specific DOM for all times.\n\n", 
    "reporter": "mzoll", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1466520708255024", 
    "component": "combo core", 
    "summary": "[steamshovel] bubbles wrong color-code", 
    "priority": "normal", 
    "keywords": "steamshovel", 
    "time": "2015-12-01T17:07:45", 
    "milestone": "Long-Term Future", 
    "owner": "hdembinski", 
    "type": "enhancement"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20151209T19:13:57: 
* hdembinski changed type from "defect" to "enhancement"
* hdembinski changed milestone from "offline V15-12-00" to "Long-Term Future"

**Comment by icecube-trac** [comment]
Trac update at 20160621T13:56:37: jvansanten commented:

> The easiest way to handle this might be to emit a new sphere with an updated, static color every time the size of the previous sphere drops to zero. This has a few advantages:
> - The current meaning of the color is preserved with the default dwell time
> - Moreover, the underlying scene is exactly the same with the default dwell time
> - No monkeying about creating and exposing new ColorVariants

**Comment by icecube-trac** [comment]
Trac update at 20160621T14:22:55: hdembinski commented:

> Replying to [comment:2 jvansanten]:
> > The easiest way to handle this might be to emit a new sphere with an updated, static color every time the size of the previous sphere drops to zero. This has a few advantages:
> > - The current meaning of the color is preserved with the default dwell time
> > - Moreover, the underlying scene is exactly the same with the default dwell time
> > - No monkeying about creating and exposing new ColorVariants
> 
> I am not sure how this could work. Steve set up the system such that the conversion of time to values happens in C++ to make the system fast. There is no callback in Python to change the scene as a function of current time. The only thing to achieve that is a hackish class called PhantomSceneObject. An invisible object in the 3D scene that allows a callback into Python code based on current time. This is used mainly to power the MPLArtists, which are themselves sort-of a hack.

**Comment by icecube-trac** [comment]
Trac update at 20160621T14:49:49: jvansanten commented:

> This would all happen at setup time as it does now. Now, it creates a single sphere:
> 
> http://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/steamshovel/trunk/python/artists/Bubbles.py#L65
> 
> and sets its size as a TimeVariant. What I suggest is that when the size function drops to zero, e.g. here:
> 
> http://code.icecube.wisc.edu/projects/icecube/browser/IceCube/projects/steamshovel/trunk/python/artists/Bubbles.py#L143
> 
> you finalize the current sphere and start a new one, with its color set to the time of the next changepoint. Then you can continue building a new TimeVariant for its size until it too vanishes, et cetera. If the size never drops to zero because the dwell time is still at its default, then nothing changes.

**Comment by icecube-trac** [comment]
Trac update at 20160621T14:51:48: 
* hdembinski changed status from "new" to "closed"
* hdembinski changed resolution from "" to "fixed"
* hdembinski commented:

> In changeset 147649:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="147649"
> fixes #1449
> ```
